### PR TITLE
ADAPT Client

### PR DESCRIPTION
**Pull request by GillySpace27**
## PR Description

Adds functionality to download ADAPT maps using FIDO.
This addresses the following issue: https://github.com/sunpy/sunpy/issues/7152

I believe that the core of this code works but I cant get a dev environment running to test things. I believe I'll have to do more work putting the right parts in the right places for sunpy, but I wanted to get this started by uploading what I have.

### CODE REVIEW

**Review by GillySpace27 — COMMENTED** [review]
Ok, I think that the client works now, though there may be some code linting that could be helpful.

File: setup.cfg
Comment (nabobalis) [inline setup.cfg:1]: So you caught sunpy during its transition to another packaging standard. 

You can delete this file!

File: sunpy/net/__init__.py
Comment (nabobalis) [inline sunpy/net/__init__.py:14]:
```suggestion
```

File: sunpy/net/dataretriever/attrs/__init__.py
Comment (nabobalis) [inline sunpy/net/dataretriever/attrs/__init__.py:8]: I do not think we need this for this file since the only two items in this file are covered by all.

File: sunpy/net/dataretriever/attrs/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/attrs/adapt.py:3]:
```suggestion
```

File: sunpy/net/dataretriever/attrs/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/attrs/adapt.py:14]: I assume this should say something else?

File: sunpy/net/dataretriever/sources/__init__.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/__init__.py:1]: I assume we have made due without this file? Can you delete this nad we can add it back if we need it?

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:3]:
```suggestion
```
The plan (but not yet formal) has been to try and not have authorship information.
But move this information into our formal documentation to highlight people who have contributed heavily to parts of sunpy.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:15]: We will need a documentation string that follows the format of the others. You can just copy and paste another one from another client and just find and replace some names.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:50]:
```suggestion
```

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:61]: I guess we want to avoid coupling coordinates and net but I think that coordinates works out of the box, so it might be safe to move these imports to the top of the file.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:74]:
```suggestion
```

File: sunpy/net/dataretriever/sources/tests/test_adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/tests/test_adapt.py:1]: I would actually suggest for the tests is to copy the tests from another client and just change the imports to the adapt client and any adapt specific attrs or values.

It will avoid you having to write the tests from scratch and it will cover the main areas we need to test for.

**Review by nabobalis — COMMENTED** [review]
Thanks for the PR @GillySpace27!

I hope it was ok for me to look over this while it was in draft.

**Review by GillySpace27 — COMMENTED** [review]
I think I did everything except for the test parts, which I'm not experienced with. Please feel free to give this another once-over.

**Comment by nabobalis** [comment]
Hi @GillySpace27, sorry for not getting back to you sooner, I will go over this soon but I have two suggestions: 

1) Can you add back "sunpy/net/dataretriever/sources/__init__.py", I think that has to be there.

2) For the tests, you can avoid having to write the tests from scratch by just copying the file from say one of the goes test files, update the client to use adapt and then tweak the calls to the client with the correct attrs. That should reduce the burden on you and would cover most of the tests we would need for a new client.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:15]:
```suggestion
    Provides access to the ADvanced Adaptive Prediction Technique (ADAPT) products of the National Solar Observatory (NSO).
```

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:17]:
```suggestion
     `Searches data hosted by the NSO <https://gong.nso.edu/adapt/maps/gong/>`__
```

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:94]: This I think should be moved elsewhere, maybe to sunpy.time? 

This also overlaps with https://docs.sunpy.org/en/stable/generated/api/sunpy.coordinates.sun.carrington_rotation_time.html

I need input from the others for the best solution.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:27]:
```suggestion
    Examples
    --------
    >>> from sunpy.net import Fido, attrs as a
    >>> from sunpy.net.datatretriver.sources.adapat import carrington_time

    >>> start_datetime, end_datetime = carrington_time(CR, frames)
    >>> longitude_type = '0'
    >>> Fido.search(attrs.Instrument('adapt'), attrs.Time(start_datetime, end_datetime), ADAPTLngType(LngType))
```

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:43]:
```suggestion
```
This will need to be replaced with just the output from the search block.
We want to avoid downloading files where possible to reduce test time and server load on providers.

**Review by GillySpace27 — COMMENTED** [review]
Thanks for the additional comments! I did everything except the tests, which I will definitely get around to ASAP. Let me know if you see anything else!

File: sunpy/net/dataretriever/sources/adapt.py
Comment (GillySpace27) [reply sunpy/net/dataretriever/sources/adapt.py:89]: Yeah, this is just a super basic wrapper around that carrington_rotation_time function, I'm glad to put it wherever (I suspect using the strftime is probably not necessary either since it is a Quantity)

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:26]:
```suggestion
    >>> res = Fido.search(a.Instrument('adapt'), a.Time(start_datetime, end_datetime), a.adapt.ADAPTLngType(longitude_type))
```

File: sunpy/net/dataretriever/sources/adapt.py
Comment (alasdairwilson) [reply sunpy/net/dataretriever/sources/adapt.py:89]: There are people better qualified to answer this than me but...

There are some issues with the implementation here but I think we can look past that:

So I assume you have this here because you think it will be useful to search via Carrington rotation rather than via a Timerange. As this stands I don't see this adding anything other than what exists on the function you have wrapped.

What I mean is that other than working out how much the end time should differ from the start time to get n results, a user can already convert a 2 CR numbers to times and then insert those into the attrs. Indeed if the user wants n results they can just select the first n rows from the query response before calling Fetch.  I also don't think it fits with the other clients where we have a unique helper function in a dataretrieverclient file that has to be called independently before calling a query.

Now what would be cool....

If searching by CR is desired then we could add a CarringtonRotation attr and have the can handle function check for either a.Time or a.CR and then, I don't know exactly where, but either in a pre_search_hook on the generic client or via attrwalker add_converter then we could optionally take either a CR number or a a.Time range to carry out an ADAPT search.

Now what form that new attr should take, I don't know. Your current API is start + number of observations but start+end makes a lot more sense rather than having an attr be composed of 2 different typed values. Is there a particular reason for your choices?

This would be much more teachable via an example than I think preconverrting the CR into times manually.

I wouldn't mind that being a separate follow on PR, as it isn't necessary for the client to operate, but I am not super keen on this function as it stands. I will bow down to those with greater knowledge though (and if so then we need to work on the implementation).

Keen to hear others thoughts.

File: sunpy/net/dataretriever/attrs/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/attrs/adapt.py:12]: Does the ADAPT API provide a range of possible values for any of these attributes?

**Review by GillySpace27 — COMMENTED** [review]
I just adjusted the main client in response to your comments, but I still haven't touched the tests, which seem daunting. Is there anything else I should do in addition to the tests?

**Comment by nabobalis** [comment]
> I just adjusted the main client in response to your comments, but I still haven't touched the tests, which seem daunting. Is there anything else I should do in addition to the tests?

I can have a look, but for the tests, I could just copy the tests from another client and just modify the names. That should cover most of the tests we would need.

File: sunpy/net/dataretriever/attrs/adapt.py
Comment (GillySpace27) [reply sunpy/net/dataretriever/attrs/adapt.py:12]: Hmmm, I think that's what the docstrings in this file are doing, is that not the case? All the possible values are also in the register_values method in the main file. How can I make this more compliant?

Comment (nabobalis) [reply sunpy/net/dataretriever/attrs/adapt.py:12]: It isn't about making it more compliant. For the VSO for example, we have a script that calls their API and it will update the values that are listed in "register_values". 

This avoids us having to do it manually and I was wondering if it was possible for ADAPT.

Comment (GillySpace27) [reply sunpy/net/dataretriever/attrs/adapt.py:12]: I see... I'm not sure if they have an API, as this is just a web scraper client, but I can look into it.

Comment (GillySpace27) [reply sunpy/net/dataretriever/attrs/adapt.py:12]: This is something, at least: https://gong.nso.edu/adapt/maps/adapt_filename_notes.txt

**Review by GillySpace27 — COMMENTED** [review]
> > I just adjusted the main client in response to your comments, but I still haven't touched the tests, which seem daunting. Is there anything else I should do in addition to the tests?

> I can have a look, but for the tests, I could just copy the tests from another client and just modify the names. That should cover most of the tests we would need. 

If you are able to do that, that would be rad! looking at the changes you make would probably be useful for me to see.

File: sunpy/net/dataretriever/attrs/adapt.py
Comment (nabobalis) [reply sunpy/net/dataretriever/attrs/adapt.py:12]: Thanks.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/adapt.py:81]: Is this correct?

**Comment by nabobalis** [comment]
Hi @GillySpace27, I rebased and updated the tests, they should cover what we need.

File: sunpy/net/dataretriever/sources/adapt.py
Comment (GillySpace27) [reply sunpy/net/dataretriever/sources/adapt.py:81]: I believe so, though I wrote it a while back. Does something seem off?

**Comment by nabobalis** [comment]
I think there's been a slight mixup in the merge.

**Comment by nabobalis** [comment]
Hi @GillySpace27, I would suggest that instead of trying to merge in the changes locally, that it would be better to run:

```
git remote update -p
git reset --hard origin/main
```

This will update your local repo and reset all local changes to point to the current version that is on GitHub. 

If you have any questions about git, please just let know!

File: sunpy/net/dataretriever/sources/adapt.py
Comment (nabobalis) [reply sunpy/net/dataretriever/sources/adapt.py:81]: Ah I misread, ignore me.

**Comment by GillySpace27** [comment]
Thanks for your help getting this sorted out! Let me know if you need anything from me. :)

**Comment by nabobalis** [comment]
> Thanks for your help getting this sorted out! Let me know if you need anything from me. :)

Sorry I kind of just took over. There were some parts that I had to recall how it worked for me to patch. I need to increase the developer guide with more details so its more helpful. 

If you are ok with the changes, and want to check it still works how you expect it to, that would be great. Then I can merge it.

**Review by nabobalis — APPROVED** [review]

**Comment by GillySpace27** [comment]
So what happens now?

**Comment by nabobalis** [comment]
Right now, I need to wait for another review before we should merge.

I am hoping that will come as we want to get 6.0 out.

**Comment by GillySpace27** [comment]
Thanks again for your help! I'll mark this off as a backburnered project until I hear back from you all. I hope it makes it in to the release!

**Comment by nabobalis** [comment]
Online fails are due to HEK and nothing to do with this PR.

**Comment by nabobalis** [comment]
Thanks for the PR @GillySpace27, sorry it took so long.

**Pull request merged**
